### ISSUE TITLE
builtins: `pg_get_constraintdef` broken during DROP CONSTRAINT in txn

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#75825]
This is a new failure introduced between 4d6d79d...7897f24, so my guess is 617d300 is the cause of the regression.

The following should not error:
```sql
BEGIN;
create table tbl (price int, quantity int);
alter table tbl add constraint quan_check check (quantity > 0);
alter table tbl add constraint pr_check check (price > 0);

SELECT conname, pg_get_constraintdef(c.oid) AS constraintdef, c.convalidated AS valid
            FROM pg_constraint c
            JOIN pg_class t ON c.conrelid = t.oid
            WHERE c.contype = 'c'
              AND t.relname = 'tbl'; -- this query beforehand  is required to trigger the error; without it everything is fine
alter table tbl drop constraint quan_check;
SELECT conname, pg_get_constraintdef(c.oid) AS constraintdef, c.convalidated AS valid
            FROM pg_constraint c
            JOIN pg_class t ON c.conrelid = t.oid
            WHERE c.contype = 'c'
              AND t.relname = 'tbl'; -- this fails
COMMIT;
```

example output:
```
root@127.0.0.1:26257/defaultdb> BEGIN;
create table tbl (price int, quantity int);
alter table tbl add constraint quan_check check (quantity > 0);
alter table tbl add constraint pr_check check (price > 0);

SELECT conname, pg_get_constraintdef(c.oid) AS constraintdef, c.convalidated AS valid
            FROM pg_constraint c
            JOIN pg_class t ON c.conrelid = t.oid
            WHERE c.contype = 'c'
              AND t.relname = 'tbl'; 
alter table tbl drop constraint quan_check;
SELECT conname, pg_get_constraintdef(c.oid) AS constraintdef, c.convalidated AS valid
            FROM pg_constraint c
            JOIN pg_class t ON c.conrelid = t.oid
            WHERE c.contype = 'c'
              AND t.relname = 'tbl';
COMMIT;
BEGIN


Time: 0ms total (execution 0ms / network 0ms)

CREATE TABLE


Time: 3ms total (execution 3ms / network 0ms)

ALTER TABLE


Time: 12ms total (execution 12ms / network 0ms)

ALTER TABLE


Time: 9ms total (execution 9ms / network 0ms)

   conname   |     constraintdef      | valid
-------------+------------------------+--------
  pr_check   | CHECK ((price > 0))    | true
  quan_check | CHECK ((quantity > 0)) | true
(2 rows)


Time: 11ms total (execution 11ms / network 0ms)

ALTER TABLE


Time: 2ms total (execution 2ms / network 0ms)

ERROR: pg_get_constraintdef(): unknown constraint (OID=1220880198)
SQLSTATE: 22023
ROLLBACK


Time: 2ms total (execution 1ms / network 2ms)
```

`pg_get_constraintdef`: https://github.com/cockroachdb/cockroach/blob/be1b6c4fd4778b784658459f77d28e6ed746c349/pkg/sql/sem/builtins/pg_builtins.go#L300-L319

Refs https://github.com/cockroachdb/cockroach/issues/68363#issuecomment-1027287913

**Comment by ajwerner** [comment]
In https://github.com/cockroachdb/cockroach/pull/69008 we changed where we decide to reset the complete set of descriptors. I'm not totally sure I picked up on that subtle detail. We use to do it when we did a write. The idea was that's when we'd invalidate the set. @postamar changed it to be on read, which I sort of get, but I don't totally get. 
https://github.com/cockroachdb/cockroach/blob/be1b6c4fd4778b784658459f77d28e6ed746c349/pkg/sql/catalog/descs/descriptor.go#L237

The idea being that you've got to check things out before you check them in, but I don't know, it doesn't seem to work here. I will note that new descriptors we avoid the whole descs collection. I don't know why that is.
